core/math/a_star_grid_2d.cpp

```diff
@@ -734,6 +734,102 @@ TypedArray<Vector2i> AStarGrid2D::get_id_path(const Vector2i &p_from_id, const V
 	return path;
 }
 
+Vector2 AStarGrid2D::get_flow_direction(const Vector2i &p_id) const {
+	ERR_FAIL_COND_V_MSG(dirty, Vector2(), "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_V_MSG(!is_in_boundsv(p_id), Vector2(), vformat("Can't get point's flow direction. Point %s out of bounds %s.", p_id, region));
+	return _get_point_unchecked(p_id)->flow_direction;
+}
+
+TypedArray<Vector2> AStarGrid2D::get_flow_directions(const Rect2i &p_region) const {
+	ERR_FAIL_COND_V_MSG(dirty, TypedArray<Vector2i>(), "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_V_MSG(!region.encloses(p_region), TypedArray<Vector2i>(), vformat("Can't get point's flow direction. Region %s out of bounds %s.", p_region, region));
+
+	TypedArray<Vector2> result;
+	for (int y = p_region.get_position().y; y < p_region.get_end().y; y++) {
+		for (int x = p_region.get_position().x; x < p_region.get_end().x; x++) {
+			result.push_back(_get_point_unchecked(Vector2i(x, y))->flow_direction);
+		}
+	}
+	return result;
+}
+
+void AStarGrid2D::set_point_cost(const Vector2i &p_id, int p_cost) {
+	ERR_FAIL_COND_MSG(dirty, "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_MSG(!is_in_boundsv(p_id), vformat("Can't set point's cost. Point %s out of bounds %s.", p_id, region));
+	ERR_FAIL_COND_MSG(p_cost < 1 || p_cost > 254, "Can't set point's cost. Cost must be within (1-254) range.");
+	_get_point_unchecked(p_id)->cost = static_cast<uint8_t>(p_cost);
+}
+
+int AStarGrid2D::get_point_cost(const Vector2i &p_id) const {
+	ERR_FAIL_COND_V_MSG(dirty, 0, "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_V_MSG(!is_in_boundsv(p_id), 0, vformat("Can't get point's cost. Point %s out of bounds %s.", p_id, region));
+	return static_cast<int>(_get_point_unchecked(p_id)->cost);
+}
+
+void AStarGrid2D::calculate_flow_field(const Vector2i &p_id, bool p_ignore_obstacles) {
+	ERR_FAIL_COND_MSG(dirty, "Grid is not initialized. Call the update method.");
+	ERR_FAIL_COND_MSG(!is_in_boundsv(p_id), vformat("Can't calculate flow field. Point %s out of bounds %s.", p_id, region));
+
+	// Clear previous results.
+	for (LocalVector<Point> &line : points) {
+		for (Point &p : line) {
+			p.flow_direction = Vector2();
+			p.value = UINT32_MAX;
+		}
+	}
+
+	// Calculate integration field.
+	{
+		LocalVector<Point *> open_list;
+
+		Point *target = _get_point(p_id.x, p_id.y);
+		target->value = 0;
+		open_list.push_back(target);
+
+		while (!open_list.is_empty()) {
+			Point *p = open_list[0];
+			open_list.remove_at(0);
```

Comment:
Use the last index instead ! Or remove_at will move all values every iteration of the loop !
```
Point *p = open_list[open_list.size() - 1];
open_list.remove_at(open_list.size() - 1);
```
Or use : `remove_at_unordered`

Reply:
Thanks, fixed, I need to remember that.